### ISSUE TITLE
【已解决】第五章 pytorch中常用的工具 P148 关于怎么将单通道灰度的minist手写数据集 预处理拼成4*4的网格（并不会像书中自动转化为3通道，需自己转化），预处理单通道要么归一化单通道的均值方差，要么将单通道转化为3通道，再归一化三通道

### ISSUE THREAD

**Issue by huangpan2507** [chenyuntc/pytorch-book#167]
from torchvision import models
from torch import nn
from torchvision import datasets
from torchvision import transforms
import torch
from torch.utils.data import DataLoader
from torchvision.utils import make_grid, save_image
from PIL import Image

transform = transforms.Compose([
    transforms.RandomSizedCrop(224),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    # 因为minist是单通道灰色图，因此此处要么扩展为三通道，要么只对三通道归一化
    transforms.Normalize((0.1307,), (0.3081,))
    # transforms.Lambda(lambda x: x.repeat(3, 1, 1)),
    # transforms.Normalize(mean=[0.4, 0.4, 0.4], std=[0.2, 0.2, 0.2]),

])

resnet34 = models.resnet34(pretrained=True, num_classes=1000)
resnet34.fc = nn.Linear(512, 10)

dataset = datasets.MNIST('data/', download=True, train=False,
                         transform=transform)

to_pil = transforms.ToPILImage()

'''这是将随机的3通道，64*64的显示出来
a = (torch.randn(3, 64, 64)).float()
to_pil = transforms.ToPILImage()(a * 0.2 + 0.4)
to_pil.show() 
# to_pil(torch.randn(3, 64, 64)).show()'''

dataloader = DataLoader(dataset, shuffle=True, batch_size=16)
dataiter = iter(dataloader)
img = make_grid(next(dataiter)[0], 4)      # 拼成4*4的网格图片
to_pil(img).show()

save_image(img, 'a.png')
Image.open('a.png')              # [读取保存的图片]

可以看下我的GitHub

**Comment by shazhongcheng** [comment]
请问你们谁有第十章数据集吗？可以百度云分享一下吗？感激不尽！

**Comment by eliasyin** [comment]
感谢楼主提供的两个解决思路！我来把楼主的解决方案整理一下吧
 > __核心的地方就是修改transforms.Compose的参数__
### 下面两段代码的运行结果和教程里边是一致的（具体的图片可能会不一样）
### 1、
```python
from torchvision import datasets
from torchvision import transforms as T
from torch.utils.data import DataLoader
from torchvision.utils import make_grid
to_img = T.ToPILImage()

transform = T.Compose([
    T.RandomResizedCrop(224),
    T.RandomHorizontalFlip(),
    T.ToTensor(),
    #将一个通道的图片转换为三个通道，这样归一化时可以使用三个通道的数据（三维数组）
    T.Lambda(lambda x: x.repeat(3, 1, 1)),
    T.Normalize(mean=[0.4, 0.4, 0.4], std=[0.2, 0.2, 0.2]),
])

dataset = datasets.MNIST('data/', download=True, train=False,
                        transform=transform)
dataloader = DataLoader(dataset, shuffle=True, batch_size=16)
dataiter = iter(dataloader)
img = make_grid(next(dataiter)[0], 4)
to_img(img)
```
### 2、
```python
from torchvision import datasets
from torchvision import transforms as T
from torch.utils.data import DataLoader
from torchvision.utils import make_grid
to_img = T.ToPILImage()

transform = T.Compose([
    T.RandomResizedCrop(224),
    T.RandomHorizontalFlip(),
    T.ToTensor(),
    #因为只有一个通道，所以只需给出一个通道的归一化参数
    T.Normalize((0.1307,), (0.3081,))
])

dataset = datasets.MNIST('data/', download=True, train=False,
                        transform=transform)
dataloader = DataLoader(dataset, shuffle=True, batch_size=16)
dataiter = iter(dataloader)
img = make_grid(next(dataiter)[0], 4)
to_img(img)
```

**Comment by huangpan2507** [comment]
嗯。可以的，欢迎交流
